### DISCUSSION TITLE
Dockerfile rebuilt every debug session AddDockerfile(..).WithLifetime(Persistent)

### DISCUSSION THREAD

**Discussion by Simon900225** [microsoft/aspire#6928] [Q&A] [answered]
We have a project that rarely changes but takes up to 30 seconds to start up. So when I read about WithLifetime(ContainerLifetime.Persistent) it seemed like a perfect fit. 
And it partially works as the container isn't killed when I stop the debugger. The issue is that the next time I start the project the Dockerfile is rebuilt. When it's done the name of the container isn't changed (dataclient-4fcad0f2) but the image is updated and the container restarts. 

The only documentation I could find was the following:
> Persistent
> Attempt to re-use a previously created resource (based on the container name) if one exists. Do not destroy the container on app host process shutdown.
> In the event that a container with the given name does not exist, a new container will always be created based on the current ContainerResource configuration.
> When an existing container IS found, Aspire MAY re-use it based on the following criteria:
> - If the container WAS NOT originally created by Aspire, the existing container will be re-used.
> 
> - If the container WAS originally created by Aspire, and the "ContainerResource" configuration DOES match the existing container, the existing container will be re-used.
> 
> - If the container WAS originally created by Aspire, and the "ContainerResource" configuration DOES NOT match the existing container, the existing container will be stopped and a new container created in order to apply the updated configuration.

As I understand it, I haven't changed the ContainerResource configuration and the container should be re-used. 

This is the code I use to set it up:
```
var dataClient = builder.AddDockerfile("dataclient", "..\\", "DataClient\\Dockerfile.debug")
    .WithHttpHealthCheck("/health")
    .WithEnvironment("INSTALLATION", "XXXXXX")
    .WithEnvironment("URL", "XXXXXXX")
    .WithHttpEndpoint(8060, 8060)
    .WithExternalHttpEndpoints()
    .WithLifetime(ContainerLifetime.Persistent);
```
Have I missed something?

**Comment by Simon900225** [comment]
Tested some more.
It works if i build the Dockerfile without aspire and use AddContainer instead. 
I guess this will work but all developers in the team needs to build the dockerfile manually before they can run the aspire solution in that case.

**Comment by davidfowl** [comment]
cc @danegsta I believe if you turn on verbose logging for DCP in your app settings, it'll show your why that is happening (likely because we change the image name)

**Reply by danegsta** [reply, marked as answer]
Yeah, verbose logging will report potential changes that triggered a rebuild of the container. Most likely, as David suggested, it’s that something about your Dockerfile is preventing your image builds from being cached (generally a COPY or ADD command pointing at a folder whose contents are changing between builds). Due to Docker build caching, building the same image multiple times with no changes to the Docker build context (the local folders Docker pulls file contents from) should result in the same final image ID. 

This is admittedly a bit fuzzy compared to the other persistent container rebuild triggers and is something where we should consider adding configuration to optionally skip building images completely if they already exist.

**Comment by sajmonr** [comment]
I ran into the same issue. We have a Dockerfile and even though the contents (of the Dockerfile or any other resource that it copies in) are not changing it is still restarted every time. 
I created a fresh application just to test this with the following Dockerfile:

```Dockerfile
FROM alpine:3.22

CMD ["/bin/sh", "-c", "sleep 3600"]
```

Notice that it does nothing except pull a base image and executes `sleep` to keep it running.
Then I ran it with Aspire like so:

```c#
var builder = DistributedApplication.CreateBuilder(args);

builder.AddDockerfile("test", "./")
    .WithLifetime(ContainerLifetime.Persistent)
    .WithContainerName("test");

builder.Build().Run();
```

Even with this minimal setup the container is recreated every single time I run the app. It makes me wonder if the persistent lifetime will not work with `AddDockerfile` regardless of the contents of the Dockerfile and will be built every time. Is that something that is expected?

For what it's worth, building outside of Aspire and just adding the container as mentioned by @Simon900225 preserves the container and works as expected.

Any comment on whether my assumption is correct would be appreciated.

**Reply by danegsta** [reply]
We actually have a change in Aspire main that makes us less aggressive about replacing persistent containers using Dockerfiles. Previously we were dependent on the container build cache to avoid replacing a persistent container if the Dockerfile and build context were unchanged (build caching means you should get the same image ID for a subsequent build if nothing has changed), but this was somewhat unreliable and lead to annoying behavior for common development scenarios (copying files into the container during build, but replacing them with a bind mount for development would cause us to replace the container every time the copied files were changed). 

The new model approach takes the build context path, contents of the Dockerfile, and build arguments into account when generating the lifecycle hash we use to determine if the persistent container config requires had changed materially. Now we’ll only do a new build of your image if those values have changed instead of building your Dockerfile every time the app host runs. 

If you’re able to try out a nightly Aspire build, you can confirm if our new approach resolves the issue for you.

**Reply by sajmonr** [reply]
Thanks for the quick response. Unfortunately, for the example above, the nightly build did nothing. It still recreates a new image every time even though the only thing the Dockerfile does is run alpine and sleeps. The versions that I got from the nightly feed were `9.4.0-preview.1.25305.2`.

```
<Project Sdk="Microsoft.NET.Sdk">

  <Sdk Name="Aspire.AppHost.Sdk" Version="9.4.0-preview.1.25305.2" />

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Aspire.Hosting.AppHost" Version="9.4.0-preview.1.25305.2" />
  </ItemGroup>

</Project>

```

**Reply by danegsta** [reply]
I see an edge case we missed; we should have a fix for the fix shortly.

**Reply by danegsta** [reply]
Just merged another fix to account for the configured Aspire image changing each run when building a dockerfile

**Reply by davidfowl** [reply]
Should the approaches make the band stable?

**Reply by danegsta** [reply]
Shouldn't need to change the image behavior; we still consider it if we're not building a Dockerfile (in which case it'll be stable until the user updates it). Main thing is we'll be less aggressive about replacing persistent containers just because they're built from a Dockerfile.

**Reply by sajmonr** [reply]
I tried latest build this morning `9.4.0-preview.1.25306.2` and the container is now stable between runs. Changing the contents of the Dockerfile correctly rebuilds the container on startup.

I know this is somewhat separate, however, is change to a content of any file that you copy into the container considered a rebuild? For example, if I have `startup.sh` and I copy it in Dockerfile and then I change the contents of the `startup.sh` _but not_ the Dockerfile, will I end up with the old container or a new one?

Thanks for all the work!

**Reply by danegsta** [reply]
Glad we got things working as expected! We’re not currently evaluating any copied files, only the contents of the Dockerfile itself; it’s a bit of a balancing act since volume mounting over copied files is a not uncommon development approach. We don’t want to make users jump through hoops in their Dockerfile to deal with Aspire persistent container recreation logic. 

The overall goal is to have “some” magic to recreate a container if its configuration has changed materially (and avoid the traditional 30+ minutes spent troubleshooting why something isn’t working only to realize the latest config hasn’t applied), but not so much magic that we’re annoying the user by pulling the container out from under them constantly and removing the entire benefit of marking it persistent.

**Reply by sajmonr** [reply]
That makes perfect sense. Thanks again!
